Introduce Item Lock API mainly for Client-Side inventory action prevention

yuyaprgrm opened the pull request:
## Introduction
Currently, PocketMine-MP have no API for item lock api provided via item's NBT(minecraft:item_lock). 
The purpose of this PR is to provide an API for item locking and implement logic for server-side validation

### Relevant issues
N/A

## Changes
### API changes
- add `Item::setLockMode(?ItemLockMode $lockMode) : self`
- add `Item::getLockMode() : ?ItemLockMode`
- add `ItemLockMode` enum

### Behavioural changes
N/A

## Backwards compatibility
N/A

## Follow-up
Currently, value used in NBT is hardcoded. This value may be defined in BedrockProtocol.

## Tests
Check whether it works correctly with ScriptPlugin below.
```php
<?php

namespace famima65536\ItemLockTest;

use pocketmine\event\EventPriority;
use pocketmine\event\player\PlayerJoinEvent;
use pocketmine\item\ItemLockMode;
use pocketmine\item\VanillaItems;
use pocketmine\plugin\PluginBase;

/**
 * @name ItemLockTest
 * @main famima65536\ItemLockTest\Main
 * @version 0.1.0
 * @api 4.9.0
 * @description A plugin makes sure that Item::setLockMode() and Item::getLockMode() works well.
 * @author famima65536
 */


class Main extends PluginBase{

	protected function onEnable(): void{
		$this->getServer()->getPluginManager()->registerEvent(PlayerJoinEvent::class, function(PlayerJoinEvent $event) : void{
			$event->getPlayer()->getInventory()->setContents([
				VanillaItems::STICK()->setLockMode(ItemLockMode::SLOT()),
				VanillaItems::APPLE()->setLockMode(ItemLockMode::INVENTORY()),
			]);
		}, EventPriority::NORMAL, $this);
	}
}
```
![image](https://user-images.githubusercontent.com/17157879/195265281-41fc8ad0-be80-4f6f-a4b3-6190653255ab.png)
![image](https://user-images.githubusercontent.com/17157879/195265287-c6d4aff7-5148-430e-aecb-7ccfdb6ae561.png)

yuyaprgrm committed:
Add ItemLockMode

yuyaprgrm committed:
add validation to prevent illegal inventory action

yuyaprgrm committed:
todo: define value in suitable package.

dktapps commented on src/inventory/transaction/action/SlotChangeAction.php:
```diff
@@ -76,6 +77,12 @@ public function validate(Player $source) : void{
 		if($this->targetItem->getCount() > $this->inventory->getMaxStackSize()){
 			throw new TransactionValidationException("Target item exceeds inventory max stack size");
 		}
+		if($this->targetItem->getLockMode()?->equals(ItemLockMode::IN_SLOT()) === true){
```
This check doesn't make any sense.

dktapps commented on src/inventory/transaction/action/DropItemAction.php:
```diff
@@ -45,6 +45,9 @@ public function validate(Player $source) : void{
 		}
+		if($this->targetItem->getLockMode() !== null){
+			throw new TransactionValidationException("Target item is locked in inventory or slot");
+		}
```
I don't think this check is necessary. If the origin item is actually locked in a slot, the SlotChangeAction changing it to air will be invalid anyway.

famima65536 replied:
This is necessary because without this validation, target item can be dropped when target item is locked in inventory.

dktapps commented on src/item/Item.php:
```diff
@@ -320,6 +334,12 @@ protected function deserializeCompoundTag(CompoundTag $tag) : void{
 			}
 		}
+		//TODO: define hardcoded value in BedrockProtocol or BedrockData?
+		$this->lockMode = match($tag->getByte("minecraft:item_lock", 0)){
```
This must use a constant (see the top of the file).

dktapps commented on src/item/Item.php:
```diff
@@ -320,6 +334,12 @@ protected function deserializeCompoundTag(CompoundTag $tag) : void{
+		//TODO: define hardcoded value in BedrockProtocol or BedrockData?
+		$this->lockMode = match($tag->getByte("minecraft:item_lock", 0)){
+			1 => ItemLockMode::IN_SLOT(),
+			2 => ItemLockMode::IN_INVENTORY(),
```
These hardcoded values should be constants.

dktapps commented on src/item/Item.php:
```diff
@@ -377,6 +397,15 @@ protected function serializeCompoundTag(CompoundTag $tag) : void{
 			$tag->removeTag("CanDestroy");
 		}
+		if($this->lockMode !== null){
+			$tag->setByte("minecraft:item_lock", match($this->lockMode->id()){
```
As above.

dktapps commented on src/item/Item.php:
```diff
@@ -377,6 +397,15 @@ protected function serializeCompoundTag(CompoundTag $tag) : void{
+		if($this->lockMode !== null){
+			$tag->setByte("minecraft:item_lock", match($this->lockMode->id()){
+				ItemLockMode::IN_SLOT()->id() => 1,
+				ItemLockMode::IN_INVENTORY()->id() => 2,
```
and again here

dktapps commented on src/item/ItemLockMode.php:
```diff
@@ -0,0 +1,46 @@
+	protected static function setup() : void{
+		self::registerAll(
+			new self("in_slot"),
+			new self("in_inventory")
```
These IN_ prefixes seem redundant.

dktapps reviewed (changes requested):
The server-side validation for this is incomplete. Since the current transaction system does not allow actions to be tracked when moving from one inventory to another, validation for this can't be easily implemented. This means that this effectively becomes a cosmetic-only change.

yuyaprgrm committed:
remove IN_ prefix

yuyaprgrm committed:
define constants that was hardcoded value before

yuyaprgrm commented:
Progress: this feature looks to be implementable with new inventory transaction system in current minor-next branch.
May I work this PR with rebasing and force-pushing?

dktapps commented:
The core transaction system remains unchanged, and implementing core logic in the network layer is pretty dubious

dktapps commented:
Blocked waiting on #6505

dktapps commented:
~~I had a think about this, and realized it can probably be done without a revamp of the whole system.~~

~~Basically, [this code](https://github.com/pmmp/PocketMine-MP/blob/cd2a1fdc1d2c327894b3b489937e9d6abb34e472/src/inventory/transaction/InventoryTransaction.php#L150) would have to split `$needItems` and `$haveItems` into `$needItemsByInventory` and `$haveItemsByInventory` according to which inventory they came from (if they came from a `SlotChangeAction`).~~

~~Then, [this code](https://github.com/pmmp/PocketMine-MP/blob/cd2a1fdc1d2c327894b3b489937e9d6abb34e472/src/inventory/transaction/InventoryTransaction.php#L169) would have to be adapted to only look in the subset of items that are locked to the same inventory.~~

~~This way, the transaction would only balance if the same number of inventory-locked items were consumed & created by the same inventory.~~

~~Seems doable but not for the faint of heart.~~

Never mind, seems I misunderstood the function of the inventory lock. Apparently it allows moving from a chest -> player's inventory but not the other way round. It also allows moving from cursor <-> player inventory.

dktapps committed:
Merge branch 'minor-next' into item-lock

dktapps committed:
why does this editor suck so bad

dktapps committed:
cleanup & modernize code

dktapps committed:
Server-side support for binding items to player inventory

dktapps committed:
Align slot-lock behaviour to vanilla
this behaves the same as inventory-lock up until it's placed into a slot, in that it can be taken from a chest (but not returned), and placed down into the player's inventory.
the only special behaviour of slot lock is that it can't be _removed_ from the slot.

dktapps committed:
Fix hardcoded tag name

dktapps committed:
Remove unnecessary check
transaction balancing & SlotChangeAction validation should stop this from happening anyway

dktapps committed:
better errors

dktapps committed:
Private consts

dktapps commented:
PR is basically done but I think the terminology needs to be updated. `ItemLockMode` doesn't give much away, particularly since the intended function is actually to lock items specifically into a _player's inventory_. Without improvement, people might think this feature works on general inventories, like I did.

dktapps committed:
Rework cases
